File: 2024/day01/python/solution.py
from pathlib import Path

input_text = (Path(__file__).parent.parent / "input.txt").read_text().strip()

# Parse input
lines = input_text.split("\n")


def part1():
    left_list = []
    right_list = []

    for line in lines:
        left, right = map(int, line.split())
        left_list.append(left)
        right_list.append(right)

    # Sort both lists
    left_list.sort()
    right_list.sort()

    # Calculate total distance
    total_distance = sum(abs(l - r) for l, r in zip(left_list, right_list))
    return total_distance


def part2():
    left_list = []
    right_list = []

    for line in lines:
        left, right = map(int, line.split())
        left_list.append(left)
        right_list.append(right)

    # Count occurrences in right list
    right_counts = {}
    for num in right_list:
        right_counts[num] = right_counts.get(num, 0) + 1

    # Calculate similarity score
    similarity_score = sum(num * right_counts.get(num, 0) for num in left_list)
    return similarity_score


if __name__ == "__main__":
    print(f"Part 1: {part1()}")
    print(f"Part 2: {part2()}")

File: 2024/day01/ruby/solution.rb
#!/usr/bin/env ruby

# Read input file
input_path = File.join(File.dirname(__FILE__), '..', 'input.txt')
input_text = File.read(input_path).strip

# Parse input into two lists
left_list = []
right_list = []

input_text.each_line do |line|
  left, right = line.split.map(&:to_i)
  left_list << left
  right_list << right
end

def part1(left_list, right_list)
  # Sort both lists
  sorted_left = left_list.sort
  sorted_right = right_list.sort

  # Calculate total distance
  total_distance = 0
  sorted_left.each_with_index do |left_val, i|
    total_distance += (left_val - sorted_right[i]).abs
  end

  total_distance
end

def part2(left_list, right_list)
  # Count occurrences in right list
  right_counts = Hash.new(0)
  right_list.each { |num| right_counts[num] += 1 }

  # Calculate similarity score
  similarity_score = 0
  left_list.each do |num|
    similarity_score += num * right_counts[num]
  end

  similarity_score
end

puts "Part 1: #{part1(left_list, right_list)}"
puts "Part 2: #{part2(left_list, right_list)}"
